tensorboard/plugins/graph/tf_graph_common/op.ts

```diff
@@ -14,435 +14,438 @@ limitations under the License.
-    // Function library nodes are generally for internal use.
-    if (opNode.name.search(FUNCTION_LIBRARY_NODE_PREFIX) == 0) {
-      return true;
-    }
-    // Nodes that lack op types should be ignored.
-    if (!opNode.op) {
-      return true;
-    }
-    // If assigned a device that is not TPU-related assume op is valid.
-    if (opNode.device && isNotTpuOp(opNode.device)) {
-      return true;
+      // Control flow ops.
+      'Enter',
+      'Exit',
+      'LoopCond',
+      'Merge',
+      'NextIteration',
+      'Switch',
+      // Ops below are inserted by the compiler.
+      '_Arg',
+      '_ParallelConcatUpdate',
+      '_Retval',
+      '_TPUCompile',
+      '_TPUExecute',
+      // Distributed TPU ops.
+      'TPUCompilationResult',
+      'TPUReplicatedInput',
+      'TPUReplicatedOutput',
+      'TPUReplicateMetadata',
+      // Checkpointing ops.
+      'MergeV2Checkpoints',
+      'RestoreV2',
+      'SaveV2',
+      // Miscellaneous CPU ops.
+      'Abort',
+      'Assert',
+      'Assign',
+      'Placeholder',
+      'PlaceholderV2',
+      'ShardedFilename',
+      'StringJoin',
+      'Variable',
+      'VariableV2',
+      'VarHandleOp',
+      // Summary ops.
+      'AudioSummary',
+      'AudioSummaryV2',
+      'DebugNumericSummary',
+      'HistogramSummary',
+      'ImageSummary',
+      'MergeSummary',
+      'ScalarSummary',
+      'StatsAggregatorSummary',
+    ];
+
+    /**
+     * Returns true if the node's inferred device is not the TPU.
+     * Note that this is only a best-effort check.
+     */
+    isNotTpuOp(opDevice: string): boolean {
+      if (opDevice.toLowerCase().search('cpu:') != -1) {
+        return true;
+      }
+      if (opDevice.toLowerCase().search('gpu:') != -1) {
+        return true;
+      }
+      return (opDevice.toLowerCase().search('tpu') == -1);
     }
-    // If assigned to the TPU_SYSTEM device, assume op is valid.
-    if (opNode.device && opNode.device.search('TPU_SYSTEM') != -1) {
-      return true;
+    opValid(opNode: OpNode): boolean {
+      // Function library nodes are generally for internal use.
+      if (opNode.name.search(FUNCTION_LIBRARY_NODE_PREFIX) == 0) {
+        return true;
+      }
+      // Nodes that lack op types should be ignored.
+      if (!opNode.op) {
+        return true;
+      }
+      // If assigned a device that is not TPU-related assume op is valid.
+      if (opNode.device && this.isNotTpuOp(opNode.device)) {
+        return true;
+      }
+      // If assigned to the TPU_SYSTEM device, assume op is valid.
+      if (opNode.device && opNode.device.search('TPU_SYSTEM') != -1) {
+        return true;
+      }
+      return TpuCompatibilityProvider.WHITELIST.indexOf(opNode.op) != -1;
     }
-    return WHITELIST.indexOf(opNode.op) != -1;
   }
 
-  export function checkOpsForCompatibility(graph: SlimGraph) {
+  export function checkOpsForCompatibility(graph: SlimGraph, provider: CompatibilityProvider) {
+    if(provider === null) {
```

Comment:
Please add whitespace after `if` for consistency in the file.
```suggestion
    if (provider === null) {
```